docs/decisions/0039-function-call-content.md

```diff
@@ -0,0 +1,435 @@
+
+**Cons**:
+- It is not possible to communicate an exception to the connector/LLM without the additional Exception/Error property.  
+- `FunctionResult` is not {de}serializable today:
+  * The `FunctionResult.ValueType` property has a `Type` type that is not serializable by JsonSerializer by default, as it is considered dangerous.  
+  * The same applies to `KernelReturnParameterMetadata.ParameterType` and `KernelParameterMetadata.ParameterType` properties of type `Type`.  
+  * The `FunctionResult.Function` property is not deserializable and should be marked with the [JsonIgnore] attribute.  
+    * A new constructor, ctr(object? value = null, IReadOnlyDictionary<string, object?>? metadata = null), needs to be added for deserialization. 
+    * The `FunctionResult.Function` property has to be nullable. It can be a breaking change? for the function filter users because the filters use `FunctionFilterContext` class that expose an instance of kernel function via the `Function` property.
+
+### Option 3.4 - FunctionResult: KernelContent
+Note: This option was suggested during a second round of review of this ADR.
+   
+This option suggests making the `FunctionResult` class a derivative of the `KernelContent` class:
+```csharp
+public class FunctionResult : KernelContent
+{
+    ....
+}
+```
+So, instead of having a separate `FunctionResultContent` class to represent the function result content, the `FunctionResult` class will inherit from the `KernelContent` class, becoming the content itself. As a result, the function result returned by the `KernelFunction.InvokeAsync` method can be directly added to the `ChatMessageContent.Items` collection:
+```csharp
+foreach (FunctionCallContent functionCall in functionCalls)
+{
+    FunctionResult result = await functionCall.InvokeAsync(kernel);
+
+    chatHistory.Add(new ChatMessageContent(AuthorRole.Tool, new ChatMessageContentItemCollection { result }));
+    // instead of
+    chatHistory.Add(new ChatMessageContent(AuthorRole.Tool, new ChatMessageContentItemCollection { new FunctionResultContent(functionCall, result) }));
+    
+    // of cause, the syntax can be simplified by having additional instance/extension methods
+    chatHistory.AddFunctionResultMessage(result); // Using the new AddFunctionResultMessage extension method of ChatHistory class
+}
+```
+
+Questions:
+- How to pass the original `FunctionCallContent` to connectors along with the function result. It's actually not clear atm whether it's needed or not. The current rationale is that some models might expect properties of the original function call, such as arguments, to be passed back to the LLM along with the function result. An argument can be made that the original function call can be found in the chat history by the connector if needed. However, a counterargument is that it may not always be possible because the chat history might be truncated to save tokens, reduce hallucination, etc.
+- How to pass function id to connector?
```

Comment:
Couple of options from my side...
- Introduce a `public class FunctionCallResultContent : FunctionResult` that includes the ID (and perhaps the original `FunctionCallContent` too). This would be returned by the `functionCall.InvokeAsync` method.
- Introduce an optional `ID` property on `FunctionResult` (not the biggest fan of this...)

Reply:
At the moment `FunctionResult` is not {de}serializable. Let me prototype to see what it would take to make it {de}serialization friendly.

Reply:
Here are the results https://github.com/microsoft/semantic-kernel/pull/5703 of the exercise to make `FunctionResult` {de}serialization friendly:  
   
- The `FunctionResult.Function` property is lost (the property value is null) during the serialization/deserialization process, as it is not possible to serialize executable code. We can try to preserve the function parameters/return type schema, but it may require custom serialization, and it's not clear if it's worth it.  
   
- The `FunctionResult.Function` property had to be changed to nullable; otherwise, `FunctionResult` deserialization fails because the property was originally serialized as null and serializer can't create function instance from null. As a result, the change propagated to the `FunctionFilterContext.Function` property, which also became nullable. This may be considered a breaking change since filter code did not expect the `Function` property to be null, and suddenly after the SK update, it's null. However, I don't see it having a null value, since the filters are executed on a valid function before the function invocation returns a result that might get serialized if added to chat history, and only after deserialization will the `FunctionResult.Function` property become null.  
   
- The `FunctionResult.Value` loses its original type during the deserialization process. So, in [the unit test](https://github.com/microsoft/semantic-kernel/pull/5703/files#r1544994422), the Guid type turned into a JsonElement because the type of the value property is an object, and the best the JsonDeserializer can do in the absence of a more specific target type is to deserialize the JSON to `JsonElement`.  
   
In summary, yes, `FunctionResult` can be serializable/deserializable and, correspondingly, can become a derivative of the `KernelContent` class and can be used as a base class - `public class FunctionCallResultContent : FunctionResult`. 

However, the cost of the change would be the loss of the `Function` property (which might not be needed by connectors, and if needed it can be found via `kernel.Plugins` connectors have access to) and unnecessary propagation of the nullable nature of the `FunctionResult.Function` property to the `FunctionFilterContext.Function` property.

Next steps to prototype:
- Inherit the `FunctionResult` from `KernelContent`.
- Inherit the `Function{Call}ResultContent` from `FunctionResult`.